### PR TITLE
[PERF] point_of_sale: improve related_models performance

### PR DESCRIPTION
**Pull request by davidmonnom**
This commit improves the performance of the related_models.js file by
using Map instead of Object and Array to keep ordering of records.

This commit also improves the performance of the product_screen.js file
by filtering product after slicing the product list.

### CODE REVIEW

File: addons/point_of_sale/static/src/app/store/models.js [hunk 1]
```diff
@@ -858,8 +857,11 @@ export class Orderline extends PosModel {
         return Boolean(this.combo_parent_id || this.combo_line_ids?.length);
     }
     findAttribute(values, customAttributes) {
+        if (values.length === 0 && customAttributes.length === 0) {
+            return [];
+        }
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/store/models.js:862]: What's the point of this change?

**Comment by caburj** [comment]
I pushed a suggestion so that you don't need to change the signature of `constructFullProductName`. Feel free to squash it.

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 2]
```diff
@@ -492,10 +495,7 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
             }
         }
 
-        orderedRecords[model] = orderedRecords[model].filter((rec) => rec.id !== record.id);
-
-        delete records[model][id];
-
+        records[model].delete(id);
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:498]:
```suggestion
        delete records[model][id];
```

`records[model]` is still an object, not a map.

Comment (caburj) [reply addons/point_of_sale/static/src/app/models/related_models.js:498]: oh nvm

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 3]
```diff
@@ -564,11 +567,12 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
+
+                if (X2MANY_TYPES.has(fields[key].type)) {
+                    return Array.from(this.indexedRecords[model][key][val]?.values() || []);
+                } else {
+                    return this.indexedRecords[model][key][val];
                 }
-                return result;
             },
             readAll() {
                 return this.orderedRecords[model];
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:578]: shouldn't this return `this.orderedRecords`?

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 4]
```diff
@@ -786,36 +814,10 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
-                }
-                orderedRecords[model].unshift(...valuesToAdd);
-            }
-
-            const event = valuesToAdd.length > 0 ? "create" : "update";
-            models[model].triggerEvents(event, values);
+        for (const [model, values] of Object.entries(callbacks)) {
+            const modelInst = models[model];
+            values.created.forEach((values) => modelInst.triggerEvents("create", values));
+            values.updated.forEach((values) => modelInst.triggerEvents("update", values));
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:820]: this is a significant change, is it fixing something else?

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 5]
```diff
@@ -564,11 +567,12 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
                     throw new Error(`Unable to get record by '${key}'`);
                 }
-                const result = this.indexedRecords[model][key][val];
-                if (result instanceof Map) {
-                    return Array.from(result.values());
+
+                if (X2MANY_TYPES.has(fields[key].type)) {
+                    return Array.from(this.indexedRecords[model][key][val]?.values() || []);
+                } else {
+                    return this.indexedRecords[model][key][val];
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:574]: Do you really need to make this change? I don't see any link to the `orderedRecords`. Can you tell me how it's linked?

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 6]
```diff
@@ -599,52 +616,54 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
-            getAllBy() {
+            getAllBy(key) {
                 return this.readAllBy(...arguments);
             },
             getAll() {
-                return this.readAll(...arguments);
+                return this.orderedRecords;
             },
-            getBy() {
+            getBy(key, val) {
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:625]: this is an unnecessary change

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 7]
```diff
@@ -599,52 +616,54 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
             },
             get() {
                 return this.read(...arguments);
             },
             getFirst() {
-                return this.readFirst(...arguments);
+                if (!(model in this.records)) {
+                    return;
+                }
+                return this.orderedRecords[0];
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:635]: change `readFirst` instead.

File: addons/point_of_sale/static/src/app/store/models.js (hunk 1 above)
Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/store/models.js:864]: if there is nothing selected in values and customattributes it will loop on product attr for nothing

File: addons/point_of_sale/static/src/app/models/related_models.js (hunk 3 above)
Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/models/related_models.js:578]: `[model]` index is directly accessed from the getter

File: addons/point_of_sale/static/src/app/models/related_models.js (hunk 4 above)
Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/models/related_models.js:820]: before this callback was call during the construction of orderedRecords object who's do not exist anymore, so we are doing this here

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 8]
```diff
@@ -577,7 +580,20 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
+                } else {
+                    const vals = {};
+                    for (const [k, v] of Object.entries(this.indexedRecords[model][key])) {
+                        if (!vals[k]) {
+                            vals[k] = [];
+                        }
+
+                        vals[k].push(...Array.from(v.values()));
+                    }
+                    return vals;
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:595]:
```suggestion
                    return mapObj(this.indexedRecords[model][key], (_, v) => Array.from(v.values()))
```

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 9]
```diff
@@ -599,12 +615,11 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
                 return result;
             },
-            // aliases
-            getAllBy() {
+            getAllBy(key) {
                 return this.readAllBy(...arguments);
             },
             getAll() {
-                return this.readAll(...arguments);
+                return this.orderedRecords;
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:622]: you forgot to adapt `readAll`. It should also return `this.orderedRecords`.

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 10]
```diff
@@ -683,6 +692,10 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
     function loadData(rawData, load = []) {
         const results = {};
         const missingRecords = {};
+        const callbacks = mapObj(processedModelDefs, () => ({
+            updated: new Map(),
+            created: new Map(),
+        }));
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:698]: There is already an array of "callbacks" per model in this file. Now, here, you are introducing a new local variable the masks the original. Why don't you use the old one?

Comment (caburj) [reply addons/point_of_sale/static/src/app/models/related_models.js:698]: I'm referring to this: https://github.com/odoo/odoo/blob/bb92c588ccd06e4f80daf4878cdfa0799cb1ab35/addons/point_of_sale/static/src/app/models/related_models.js#L182

Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/models/related_models.js:698]: Hello that's normal, isn't the same use :)

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 11]
```diff
@@ -786,36 +804,10 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {})
...
-                }
-                orderedRecords[model].unshift(...valuesToAdd);
-            }
-
-            const event = valuesToAdd.length > 0 ? "create" : "update";
-            models[model].triggerEvents(event, values);
+        for (const [model, values] of Object.entries(callbacks)) {
+            const modelInst = models[model];
+            values.created.forEach((values) => modelInst.triggerEvents("create", values));
+            values.updated.forEach((values) => modelInst.triggerEvents("update", values));
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:810]: `(values) => modelInst.triggerEvents("create", values)`

`values` in that callback isn't really an array of value, it's just single value.

Comment (caburj) [reply addons/point_of_sale/static/src/app/models/related_models.js:810]: And the `callback` that is called by `triggerEvents` expects an array.

File: addons/point_of_sale/static/src/app/screens/product_screen/product_screen.js [hunk 12]
```diff
@@ -341,10 +348,7 @@ export class ProductScreen extends Component {
             list = this.pos.models["product.product"].getAll();
         }
 
-        list = list
-            .filter((product) => !this.getProductListToNotDisplay().includes(product.id))
-            .slice(0, 100);
-
+        list = list.slice(0, 100);
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/product_screen/product_screen.js:351]:
```suggestion
        const productsToNotDisplay = this.getProductListToNotDisplay();
        list = list
            .filter((product) => !productsToNotDisplay.has(product.id))
            .slice(0, 100);
```

File: addons/point_of_sale/static/src/app/screens/product_screen/product_screen.xml [hunk 13]
```diff
@@ -26,8 +26,10 @@
                 <div t-ref="products" class="overflow-y-auto flex-grow-1">
                     <t t-if="!ui.isSmall or getChildCategories(this.pos.selectedCategory).length === 0 or searchWord or pos.showResultMobile">
                         <div t-if="productsToDisplay.length != 0 and pos.posHasValidProduct()" t-attf-class="product-list {{this.pos.productListViewMode}} p-1">
+                            <t t-set="productToNotDisplay" t-value="this.getProductListToNotDisplay()"/>
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/product_screen/product_screen.xml:29]:
```suggestion
```

File: addons/point_of_sale/static/src/app/screens/product_screen/product_screen.xml [hunk 14]
```diff
@@ -26,8 +26,10 @@
...
                             <t  t-foreach="productsToDisplay" t-as="product" t-key="product.id">
                                 <ProductCard
+                                    t-if="!productToNotDisplay.has(product.id)"
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/product_screen/product_screen.xml:32]:
```suggestion
```

**Review by caburj — CHANGES_REQUESTED** [review]
I just notice this. We can avoid changing the template if we do the logic js.

**Review by caburj — APPROVED** [review]
robodoo r+

**Comment by caburj** [comment]
robodoo r+

**Comment by caburj** [comment]
robodoo retry
